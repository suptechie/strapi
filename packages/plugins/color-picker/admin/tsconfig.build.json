{
  "extends": "./admin/tsconfig.json",
  "include": ["./admin/src"],
  "exclude": ["**/*.test.tsx"],
  "compilerOptions": {
    "outDir": "./dist/admin"
  }
}
