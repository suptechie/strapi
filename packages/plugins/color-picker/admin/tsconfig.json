{
  "extends": "tsconfig/base.json",
  "compilerOptions": {
    "jsx": "react-jsx",
    "noEmit": true
  },
  "include": ["src"],
  "exclude": ["node_modules"]
}
