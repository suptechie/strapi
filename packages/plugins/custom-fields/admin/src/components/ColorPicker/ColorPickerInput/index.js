import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Stack } from '@strapi/design-system/Stack';
import { Typography } from '@strapi/design-system/Typography';
import { Flex } from '@strapi/design-system/Flex';
import { Box } from '@strapi/design-system/Box';
import { BaseButton } from '@strapi/design-system/BaseButton';
import { Field, FieldHint, FieldError, FieldLabel, FieldInput } from '@strapi/design-system/Field';
import CarretDown from '@strapi/icons/CarretDown';
import { useIntl } from 'react-intl';
import { HexColorPicker } from 'react-colorful';
import useOnClickOutside from '../../../hooks/useOnClickOutside';
import getTrad from '../../../utils/getTrad';

const ColorPreview = styled.div`
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: ${(props) => props.color};
  border: 1px solid rgba(0, 0, 0, 0.1);
`;

const ColorPicker = styled(HexColorPicker)`
  && {
    width: 100%;
    height: 152px;
  }

  .react-colorful__pointer {
    width: ${({ theme }) => theme.spaces[3]};
    height: ${({ theme }) => theme.spaces[3]};
  }

  .react-colorful__saturation {
    border-radius: ${({ theme }) => theme.spaces[1]};
    margin-bottom: ${({ theme }) => theme.spaces[2]};
  }

  .react-colorful__hue {
    border-radius: 10px;
    height: ${({ theme }) => theme.spaces[3]};
  }
`;

const ColorPickerToggle = styled(BaseButton)`
  display: flex;
  justify-content: space-between;
  align-items: center;

  svg {
    width: ${({ theme }) => theme.spaces[2]};
    height: ${({ theme }) => theme.spaces[2]};
  }

  svg > path {
    fill: ${({ theme }) => theme.colors.neutral500};
    justify-self: flex-end;
  }
`;

const ColorPickerInput = ({
  attribute,
  description,
  disabled,
  error,
  intlLabel,
  labelAction,
  name,
  onChange,
  required,
  value,
}) => {
  const [showColorPicker, setShowColorPicker] = useState(false);
  const colorPickerRef = useRef();
  const colorPickerButtonRef = useRef();
  const { formatMessage } = useIntl();
  const color = value || '#ffffff';
  const styleUpcase = { textTransform: 'uppercase' };

  const listener = (event) => {
    if (
      !colorPickerRef.current ||
      colorPickerRef.current.contains(event.target) ||
      colorPickerButtonRef.current.contains(event.target)
    ) {
      // Do nothing if the click involves color picker refs
      return;
    }

    setShowColorPicker(false);
  };

  useOnClickOutside(colorPickerRef, listener);

  return (
    <Field
      name={name}
      id={name}
      // GenericInput calls formatMessage and returns a string for the error
      error={error}
      hint={description && formatMessage(description)}
    >
      <Stack spacing={1}>
        <Flex>
          <FieldLabel required={required}>{formatMessage(intlLabel)}</FieldLabel>
          {labelAction && <Box paddingLeft={1}>{labelAction}</Box>}
        </Flex>
        <ColorPickerToggle
          ref={colorPickerButtonRef}
          disabled={disabled}
          onClick={() => setShowColorPicker(!showColorPicker)}
        >
          <Flex>
            <ColorPreview color={color} />
            <Typography style={styleUpcase} variant="omega">
              {color}
            </Typography>
          </Flex>
          <CarretDown />
        </ColorPickerToggle>
        {showColorPicker && (
          <Box
            ref={colorPickerRef}
            width="240px"
            padding={2}
            background="neutral0"
            borderColor="neutral200"
            shadow="filterShadow"
            hasRadius
          >
            <ColorPicker
              color={color}
              onChange={(hexValue) =>
                onChange({ target: { name, value: hexValue, type: attribute.type } })
              }
            />
            <Flex paddingTop={2} paddingLeft={4} justifyContent="flex-end">
              <Box paddingRight={2}>
                <Typography variant="omega">
                  {formatMessage({ id: getTrad('input.format'), defaultMessage: 'HEX' })}
                </Typography>
              </Box>
              <FieldInput style={styleUpcase} value={color} onChange={onChange} />
            </Flex>
          </Box>
        )}
        <FieldHint />
        <FieldError />
      </Stack>
    </Field>
  );
};

ColorPickerInput.defaultProps = {
  description: null,
  disabled: false,
  error: null,
  labelAction: null,
  required: false,
  value: '',
};

ColorPickerInput.propTypes = {
  intlLabel: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  attribute: PropTypes.object.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.object,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  labelAction: PropTypes.object,
  required: PropTypes.bool,
  value: PropTypes.string,
};

export default ColorPickerInput;
