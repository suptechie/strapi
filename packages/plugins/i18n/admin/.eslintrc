{
  "root": true,
  "extends": ["custom/front/typescript"],
  "rules": {
    "check-file/filename-naming-convention": "off",
    "check-file/no-index": "off",
    "import/no-default-export": "off"
  },
  "overrides": [
    {
      "files": ["./tests/*", "**/*.test.*"],
      "env": {
        "jest": true
      },
      "rules": {
        /**
         * So we can do `import { render } from '@tests/utils'`
         */
        "import/no-unresolved": "off"
      }
    }
  ]
}
