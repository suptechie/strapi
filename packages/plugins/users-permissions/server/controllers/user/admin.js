'use strict';

const _ = require('lodash');
const { contentTypes: contentTypesUtils } = require('@strapi/utils');
const { ApplicationError, ValidationError } = require('@strapi/utils').errors;
const { getService } = require('../../utils');
const { validateCreateUserBody, validateUpdateUserBody } = require('../validation/user');

const { UPDATED_BY_ATTRIBUTE, CREATED_BY_ATTRIBUTE } = contentTypesUtils.constants;

const userModel = 'plugin::users-permissions.user';
const ACTIONS = {
  read: 'plugin::content-manager.explorer.read',
  create: 'plugin::content-manager.explorer.create',
  edit: 'plugin::content-manager.explorer.update',
  delete: 'plugin::content-manager.explorer.delete',
};

class NotFoundError extends Error {}
class ForbiddenError extends Error {}

const findEntityAndCheckPermissions = async (ability, action, model, id) => {
  const entity = await strapi.query('plugin::users-permissions.user').findOne({ where: { id } });

  if (_.isNil(entity)) {
    throw new NotFoundError();
  }

  const pm = strapi.admin.services.permission.createPermissionsManager({ ability, action, model });

  const roles = _.has(entity, `${CREATED_BY_ATTRIBUTE}.id`)
    ? await strapi.query('admin::role').findMany({
        where: {
          users: { id: entity[CREATED_BY_ATTRIBUTE].id },
        },
      })
    : [];

  const entityWithRoles = _.set(_.cloneDeep(entity), `${CREATED_BY_ATTRIBUTE}.roles`, roles);

  if (pm.ability.cannot(pm.action, pm.toSubject(entityWithRoles))) {
    throw new ForbiddenError();
  }

  return { pm, entity };
};

module.exports = {
  /**
   * Create a/an user record.
   * @return {Object}
   */
  async create(ctx) {
    const { body } = ctx.request;
    const { user: admin, userAbility } = ctx.state;

    const { email, username } = body;

    const pm = strapi.admin.services.permission.createPermissionsManager({
      ability: userAbility,
      action: ACTIONS.create,
      model: userModel,
    });

    if (!pm.isAllowed) {
      return ctx.forbidden();
    }

    const sanitizedBody = pm.pickPermittedFieldsOf(body, { subject: userModel });

    const advanced = await strapi
      .store({ type: 'plugin', name: 'users-permissions', key: 'advanced' })
      .get();

    await validateCreateUserBody(ctx.request.body);

    const userWithSameUsername = await strapi
      .query('plugin::users-permissions.user')
      .findOne({ where: { username } });

    if (userWithSameUsername) {
      throw new ApplicationError('Username already taken');
    }

    if (advanced.unique_email) {
      const userWithSameEmail = await strapi
        .query('plugin::users-permissions.user')
        .findOne({ where: { email: email.toLowerCase() } });

      if (userWithSameEmail) {
        throw new ApplicationError('Email already taken');
      }
    }

    const user = {
      ...sanitizedBody,
      provider: 'local',
      [CREATED_BY_ATTRIBUTE]: admin.id,
      [UPDATED_BY_ATTRIBUTE]: admin.id,
    };

    user.email = _.toLower(user.email);

    if (!user.role) {
      const defaultRole = await strapi
        .query('plugin::users-permissions.role')
        .findOne({ where: { type: advanced.default_role } });

      user.role = defaultRole.id;
    }

    try {
      const data = await getService('user').add(user);

      ctx.created(pm.sanitize(data, { action: ACTIONS.read }));
    } catch (error) {
      throw new ApplicationError(error.message);
    }
  },
  /**
   * Update a/an user record.
   * @return {Object}
   */

  async update(ctx) {
    const { id } = ctx.params;
    const { body } = ctx.request;
    const { user: admin, userAbility } = ctx.state;

    const advancedConfigs = await strapi
      .store({ type: 'plugin', name: 'users-permissions', key: 'advanced' })
      .get();

    const { email, username, password } = body;

    let pm;
    let user;

    try {
      const { pm: permissionManager, entity } = await findEntityAndCheckPermissions(
        userAbility,
        ACTIONS.edit,
        userModel,
        id
      );
      pm = permissionManager;
      user = entity;
    } catch (e) {
      if (e instanceof NotFoundError) return ctx.notFound();
      if (e instanceof ForbiddenError) return ctx.forbidden();
      throw e;
    }

    await validateUpdateUserBody(ctx.request.body);

    if (_.has(body, 'password') && !password && user.provider === 'local') {
      throw new ValidationError('password.notNull');
    }

    if (_.has(body, 'username')) {
      const userWithSameUsername = await strapi
        .query('plugin::users-permissions.user')
        .findOne({ where: { username } });

      if (userWithSameUsername && userWithSameUsername.id != id) {
        throw new ApplicationError('Username already taken');
      }
    }

    if (_.has(body, 'email') && advancedConfigs.unique_email) {
      const userWithSameEmail = await strapi
        .query('plugin::users-permissions.user')
        .findOne({ where: { email: _.toLower(email) } });

      if (userWithSameEmail && userWithSameEmail.id != id) {
        throw new ApplicationError('Email already taken');
      }
      body.email = _.toLower(body.email);
    }

    const sanitizedData = pm.pickPermittedFieldsOf(body, { subject: pm.toSubject(user) });
    const updateData = _.omit({ ...sanitizedData, updatedBy: admin.id }, 'createdBy');

    const data = await getService('user').edit({ id }, updateData);

    ctx.body = pm.sanitize(data, { action: ACTIONS.read });
  },
};
