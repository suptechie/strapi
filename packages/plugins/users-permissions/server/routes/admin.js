'use strict';

module.exports = {
  routes: [
    {
      method: 'GET',
      path: '/',
      handler: 'users-permissions.index',
    },
    {
      method: 'GET',
      path: '/search/:id',
      handler: 'users-permissions.searchUsers',
    },
    {
      method: 'GET',
      path: '/policies',
      handler: 'users-permissions.getPolicies',
    },
    {
      method: 'GET',
      path: '/roles/:id',
      handler: 'users-permissions.getRole',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.roles.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/roles',
      handler: 'users-permissions.getRoles',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.roles.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/routes',
      handler: 'users-permissions.getRoutes',
    },
    {
      method: 'GET',
      path: '/email-templates',
      handler: 'users-permissions.getEmailTemplate',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.email-templates.read'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/email-templates',
      handler: 'users-permissions.updateEmailTemplate',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.email-templates.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/advanced',
      handler: 'users-permissions.getAdvancedSettings',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.advanced-settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/advanced',
      handler: 'users-permissions.updateAdvancedSettings',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.advanced-settings.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/permissions',
      handler: 'users-permissions.getPermissions',
    },
    {
      method: 'GET',
      path: '/providers',
      handler: 'users-permissions.getProviders',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.providers.read'],
            },
          },
        ],
      },
    },

    {
      method: 'PUT',
      path: '/providers',
      handler: 'users-permissions.updateProviders',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.providers.update'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/roles',
      handler: 'users-permissions.createRole',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.roles.create'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/roles/:role',
      handler: 'users-permissions.updateRole',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.roles.update'],
            },
          },
        ],
      },
    },
    {
      method: 'DELETE',
      path: '/roles/:role',
      handler: 'users-permissions.deleteRole',
      config: {
        policies: [
          {
            name: 'admin::hasPermissions',
            options: {
              actions: ['plugin::users-permissions.roles.delete'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/connect/(.*)',
      handler: 'auth.connect',
      config: {
        policies: ['plugin::users-permissions.rateLimit'],
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/auth/local',
      handler: 'auth.callback',
      config: {
        policies: ['plugin::users-permissions.rateLimit'],
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/auth/local/register',
      handler: 'auth.register',
      config: {
        policies: ['plugin::users-permissions.rateLimit'],
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/auth/:provider/callback',
      handler: 'auth.callback',
      config: {
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/auth/forgot-password',
      handler: 'auth.forgotPassword',
      config: {
        policies: ['plugin::users-permissions.rateLimit'],
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/auth/reset-password',
      handler: 'auth.resetPassword',
      config: {
        policies: ['plugin::users-permissions.rateLimit'],
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/auth/email-confirmation',
      handler: 'auth.emailConfirmation',
      config: {
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/auth/send-email-confirmation',
      handler: 'auth.sendEmailConfirmation',
      config: {
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/users/count',
      handler: 'user.count',
      config: {
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/users',
      handler: 'user.find',
      config: {
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/users/me',
      handler: 'user.me',
      config: {
        prefix: '',
      },
    },
    {
      method: 'GET',
      path: '/users/:id',
      handler: 'user.findOne',
      config: {
        prefix: '',
      },
    },
    {
      method: 'POST',
      path: '/users',
      handler: 'user.create',
      config: {
        prefix: '',
      },
    },
    {
      method: 'PUT',
      path: '/users/:id',
      handler: 'user.update',
      config: {
        prefix: '',
      },
    },
    {
      method: 'DELETE',
      path: '/users/:id',
      handler: 'user.destroy',
      config: {
        prefix: '',
      },
    },
  ],
};
