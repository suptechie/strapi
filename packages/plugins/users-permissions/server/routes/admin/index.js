'use strict';

const usersPermissionsRoutes = require('./users-permissions');
const roleRoutes = require('./role');
const userRoutes = require('./user');
const authRoutes = require('./auth');

module.exports = {
  type: 'admin',
  routes: [...roleRoutes, ...usersPermissionsRoutes, ...authRoutes, ...userRoutes],
};
