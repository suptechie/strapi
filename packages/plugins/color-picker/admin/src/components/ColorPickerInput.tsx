import * as React from 'react';

import {
  BaseButton,
  Box,
  Field,
  Flex,
  FocusTrap,
  Popover,
  Typography,
  useComposedRefs,
} from '@strapi/design-system';
import { CaretDown } from '@strapi/icons';
import { useField, type InputProps } from '@strapi/strapi/admin';
import { HexColorPicker } from 'react-colorful';
import { useIntl } from 'react-intl';
import { styled } from 'styled-components';

import { getTrad } from '../utils/getTrad';

const ColorPreview = styled.div`
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: ${(props) => props.color};
  border: 1px solid rgba(0, 0, 0, 0.1);
`;

const ColorPicker = styled(HexColorPicker)`
  && {
    width: 100%;
    aspect-ratio: 1.5;
  }

  .react-colorful__pointer {
    width: ${({ theme }) => theme.spaces[3]};
    height: ${({ theme }) => theme.spaces[3]};
  }

  .react-colorful__saturation {
    border-radius: ${({ theme }) => theme.spaces[1]};
    border-bottom: none;
  }

  .react-colorful__hue {
    border-radius: 10px;
    height: ${({ theme }) => theme.spaces[3]};
    margin-top: ${({ theme }) => theme.spaces[2]};
  }
`;

const ColorPickerToggle = styled(BaseButton)`
  display: flex;
  justify-content: space-between;
  align-items: center;

  svg {
    width: ${({ theme }) => theme.spaces[2]};
    height: ${({ theme }) => theme.spaces[2]};
  }

  svg > path {
    fill: ${({ theme }) => theme.colors.neutral500};
    justify-self: flex-end;
  }
`;

const ColorPickerPopover = styled(Popover)`
  padding: ${({ theme }) => theme.spaces[2]};
  min-height: 270px;
`;

type ColorPickerInputProps = InputProps & {
  labelAction?: React.ReactNode;
};

export const ColorPickerInput = React.forwardRef<HTMLButtonElement, ColorPickerInputProps>(
  ({ hint, disabled = false, labelAction, label, name, required = false }, forwardedRef) => {
    const { onChange, value, error } = useField(name);
    const [showColorPicker, setShowColorPicker] = React.useState(false);
    const colorPickerButtonRef = React.useRef<HTMLButtonElement>(null!);
    const { formatMessage } = useIntl();
    const color = value || '#000000';

    const handleBlur: React.FocusEventHandler<HTMLDivElement> = (e) => {
      e.preventDefault();

      if (!e.currentTarget.contains(e.relatedTarget)) {
        setShowColorPicker(false);
      }
    };

    const composedRefs = useComposedRefs(forwardedRef, colorPickerButtonRef);

    return (
      <Field.Root name={name} id={name} error={error} hint={hint} required={required}>
        <Flex direction="column" alignItems="stretch" gap={1}>
          <Field.Label action={labelAction}>{label}</Field.Label>
          <ColorPickerToggle
            ref={composedRefs}
            aria-label={formatMessage({
              id: getTrad('color-picker.toggle.aria-label'),
              defaultMessage: 'Color picker toggle',
            })}
            aria-controls="color-picker-value"
            aria-haspopup="dialog"
            aria-expanded={showColorPicker}
            aria-disabled={disabled}
            disabled={disabled}
            onClick={() => setShowColorPicker(!showColorPicker)}
          >
            <Flex>
              <ColorPreview color={color} />
              <Typography
                style={{ textTransform: 'uppercase' }}
                textColor={value ? undefined : 'neutral600'}
                variant="omega"
              >
                {color}
              </Typography>
            </Flex>
            <CaretDown aria-hidden />
          </ColorPickerToggle>
          {showColorPicker && (
            <ColorPickerPopover
              onBlur={handleBlur}
              role="dialog"
              source={colorPickerButtonRef}
              spacing={4}
            >
              <FocusTrap onEscape={() => setShowColorPicker(false)}>
                <ColorPicker color={color} onChange={(hexValue) => onChange(name, hexValue)} />
                <Flex paddingTop={3} paddingLeft={4} justifyContent="flex-end">
                  <Box paddingRight={2}>
                    <Typography variant="omega" tag="label" textColor="neutral600">
                      {formatMessage({
                        id: getTrad('color-picker.input.format'),
                        defaultMessage: 'HEX',
                      })}
                    </Typography>
                  </Box>
                  <Field.Root>
                    <Field.Input
                      aria-label={formatMessage({
                        id: getTrad('color-picker.input.aria-label'),
                        defaultMessage: 'Color picker input',
                      })}
                      style={{ textTransform: 'uppercase' }}
                      value={value}
                      placeholder="#000000"
                      onChange={onChange}
                    />
                  </Field.Root>
                </Flex>
              </FocusTrap>
            </ColorPickerPopover>
          )}
          <Field.Hint />
          <Field.Error />
        </Flex>
      </Field.Root>
    );
  }
);
