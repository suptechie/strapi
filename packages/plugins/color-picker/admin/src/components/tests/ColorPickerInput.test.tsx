import { DesignSystemProvider } from '@strapi/design-system';
import { Form } from '@strapi/strapi/admin';
import { render as renderRTL } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';

import { ColorPickerInput } from '../ColorPickerInput';

const render = () => ({
  ...renderRTL(
    <ColorPickerInput
      name="color"
      label={'color-picker'}
      type="string"
      initialValue=""
      value=""
      onChange={() => {}}
    />,
    {
      wrapper: ({ children }) => {
        const locale = 'en';
        return (
          <IntlProvider locale={locale} messages={{}} textComponent="span">
            <DesignSystemProvider locale={locale}>
              <Form onSubmit={jest.fn()} method="POST">
                {children}
              </Form>
            </DesignSystemProvider>
          </IntlProvider>
        );
      },
    }
  ),
  user: userEvent.setup(),
});

describe('<ColorPickerInput />', () => {
  it('renders and matches the snapshot', () => {
    const { container } = render();

    expect(container).toMatchSnapshot();
  });

  it('toggles the popover', async () => {
    const { user, getByRole } = render();
    await user.click(getByRole('button', { name: 'Color picker toggle' }));

    expect(getByRole('dialog')).toBeVisible();
    expect(getByRole('slider', { name: 'Color' })).toBeVisible();
    expect(getByRole('slider', { name: 'Hue' })).toBeVisible();
    expect(getByRole('textbox', { name: 'Color picker input' })).toBeVisible();
  });
});
