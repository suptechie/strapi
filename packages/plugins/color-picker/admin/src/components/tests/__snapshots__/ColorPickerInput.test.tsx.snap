// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPickerInput /> renders and matches the snapshot 1`] = `
.c1 {
  display: flex;
}

.c3 {
  background: #ffffff;
  padding-block: 8px;
  padding-inline: 8px;
  border-radius: 4px;
  border-color: #dcdce4;
  border: 1px solid #dcdce4;
  cursor: pointer;
}

.c2 {
  font-size: 1.2rem;
  line-height: 1.33;
  font-weight: 600;
  color: #32324d;
}

.c8 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: #666687;
}

.c0 {
  align-items: stretch;
  display: flex;
  flex-direction: column;
  gap: 4px;
}

.c4 {
  align-items: center;
  display: flex;
  flex-direction: row;
}

.c9 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c5 {
  position: relative;
  outline: none;
}

.c5[aria-disabled='true'] {
  pointer-events: none;
}

.c5:after {
  transition-property: all;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c5:focus-visible {
  outline: none;
}

.c5:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

.c7 {
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: #000000;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.c6 {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.c6 svg {
  width: 8px;
  height: 8px;
}

.c6 svg>path {
  fill: #8e8ea9;
  justify-self: flex-end;
}

<div>
  <form
    method="POST"
    novalidate=""
  >
    <div
      class="c0"
    >
      <div
        class="c0"
      >
        <label
          alignitems="center"
          class="c1 c2"
          for="color"
          id="color-label"
        >
          color-picker
        </label>
        <button
          aria-controls="color-picker-value"
          aria-disabled="false"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-label="Color picker toggle"
          class="c3 c4 c5 c6"
          type="button"
        >
          <div
            class="c4"
          >
            <div
              class="c7"
              color="#000000"
            />
            <span
              class="c8"
              style="text-transform: uppercase;"
            >
              #000000
            </span>
          </div>
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="16"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m26.708 12.708-10 10a1 1 0 0 1-1.415 0l-10-10A1 1 0 0 1 6 11h20a1 1 0 0 1 .707 1.707"
            />
          </svg>
        </button>
      </div>
    </div>
  </form>
  <span
    class="c9"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
